### PR TITLE
refactor energy / require C++14 in cmake / cosmetics

### PR DESCRIPTION
**Pull request by simonpintarelli**
refactor functions to compute total energy

- Move functions to compute the total energy to `src/energy.cpp`.
- update `sirius_api.cpp`
- remove `DFT_ground_state::energy_kin_sum_pw`, which is not used anywhere.
- const correctness in K_point_set

cmake: switch to C++14

In `sirius_api.cpp`:
- replace macros GET_GS, GET_SIM_CTX by functions

The following member functions have been remove from `DFT_ground_state`
- `ewald_energy`
- `energy_{vxc,exc,vha,bxc,enuc,veff,kin}`
- `core_eval_sum`

`DFT_ground_state::total_energy` will now call stand-alone `sirius::total_energy`.

The ewald energy is still computed and stored in the `DFT_ground_state` class.

### CODE REVIEW

**Comment by toxa81** [comment]
@simonpintarelli  I also think that in  DFT_ground_state() constructor it should be sirius::ewald_energy(), otherwise it computes this->ewald_energy(), which is not what you want.

File: src/energy.hpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
+#ifndef ENERGY_H
```
Comment (toxa81) [inline src/energy.hpp]: __ENERGY_HPP__

File: src/sirius_api.cpp [hunk 2]
```diff
@@ -31,8 +31,11 @@
 #define GET_SIM_CTX(h) assert(h != nullptr); \
                        auto& sim_ctx = static_cast<utils::any_ptr*>(*h)->get<sirius::Simulation_context>();
 
-#define GET_GS(h) assert(h != nullptr); \
-                  auto& gs = static_cast<utils::any_ptr*>(*h)->get<sirius::DFT_ground_state>();
+sirius::DFT_ground_state& get_gs(void* const* h)
```
Comment (toxa81) [inline src/sirius_api.cpp]: This is good!

File: src/energy.hpp (hunk 1 above)
Comment (toxa81) [reply src/energy.hpp]: `__ENERGY_HPP__`

File: src/Potential/xc_functional.hpp [hunk 3]
```diff
@@ -113,7 +113,6 @@ namespace sirius {
                                     v1[0], v1[1], v1[2],
                                     v2[0], v2[1], v2[2],
                                     v3[0], v3[1], v3[2]);
-                MPI_Comm fft_comm = fft__.communicator();
```
[the remark this answers is not in the document]
Comment (simonpintarelli) [reply src/Potential/xc_functional.hpp]: It is called explicityl a few lines below, L120: `vdwxc_init_mpi(handler_vdw_, fft__.communicator());`

File: src/energy.hpp (hunk 1 above)
Comment (simonpintarelli) [reply src/energy.hpp]: fixed

**Pull request merged**